iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorQueryCompactor.java

```diff
@@ -44,22 +55,51 @@ public boolean run(CompactorContext context) throws IOException, HiveException,
     Map<String, String> tblProperties = context.getTable().getParameters();
     LOG.debug("Initiating compaction for the {} table", compactTableName);
 
-    String compactionQuery = String.format("insert overwrite table %s select * from %<s",
-        compactTableName);
+    HiveConf conf = new HiveConf(context.getConf());
+    String partSpec = context.getCompactionInfo().partName;
+    String compactionQuery;
+
+    if (partSpec == null) {
+      HiveConf.setVar(conf, ConfVars.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());
+      compactionQuery = String.format("insert overwrite table %s select * from %<s", compactTableName);
+    } else {
+      Hive db = Hive.get(conf);
+      org.apache.hadoop.hive.ql.metadata.Table table = db.getTable(context.getTable().getDbName(),
+          context.getTable().getTableName());
+      Map<String, String> partSpecMap = new LinkedHashMap<>();
+      Warehouse.makeSpecFromName(partSpecMap, new Path(partSpec), null);
+
+      List<FieldSchema> partitionKeys = table.getStorageHandler().getPartitionKeys(table);
+      List<String> partValues = partitionKeys.stream().map(
+          fs -> String.join("=", HiveUtils.unparseIdentifier(fs.getName()),
+              TypeInfoUtils.convertStringToLiteralForSQL(partSpecMap.get(fs.getName()),
+                  ((PrimitiveTypeInfo) TypeInfoUtils.getTypeInfoFromTypeString(fs.getType())).getPrimitiveCategory())
+          )
+      ).collect(Collectors.toList());
+
+      String queryFields = table.getCols().stream()
+          .map(FieldSchema::getName)
+          .filter(col -> !partSpecMap.containsKey(col))
+          .collect(Collectors.joining(","));
+
+      compactionQuery = String.format("insert overwrite table %1$s partition(%2$s) select %4$s from %1$s where %3$s",
+          compactTableName,
+          StringUtils.join(partValues, ","),
+          StringUtils.join(partValues, " and "),
+          queryFields);
+    }
+
+    SessionState sessionState = setupQueryCompactionSession(conf, context.getCompactionInfo(), tblProperties);
 
-    SessionState sessionState = setupQueryCompactionSession(context.getConf(),
-        context.getCompactionInfo(), tblProperties);
-    HiveConf.setVar(context.getConf(), ConfVars.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());
     try {
-      DriverUtils.runOnDriver(context.getConf(), sessionState, compactionQuery);
+      DriverUtils.runOnDriver(conf, sessionState, compactionQuery);
       LOG.info("Completed compaction for table {}", compactTableName);
+      return true;
     } catch (HiveException e) {
-      LOG.error("Error doing query based {} compaction", RewritePolicy.ALL_PARTITIONS.name(), e);
-      throw new RuntimeException(e);
+      LOG.error("Failed compacting table {}", compactTableName, e);
```

Comment:
"Failed to compact table {}, partition {} "

Reply:
Done